### PR TITLE
Ability to send custom sms using Twilio during OTP authentication

### PR DESCRIPTION
**Pull request by aseem-hegshetye**
Currently the default sms thats sent by using Twilio during 2FA is -
`Your authentication token is %s`

With this PR we can customize the message by setting variable `settings.TWILIO_MESSAGE`
Example:
`settings.TWILIO_MESSAGE='Your authentication token is {}'`

### CODE REVIEW

File: tests/test_gateways.py [hunk 1]
```diff
@@ -59,8 +61,12 @@ def test_gateway(self, client):
                 url='http://testserver/twilio/inbound/two_factor/%s/?locale=en-us' % code)
 
             twilio.send_sms(device=Mock(number=PhoneNumber.from_string('+123')), token=code)
+
             client.return_value.messages.create.assert_called_with(
-                to='+123', body='Your authentication token is %s' % code, from_='+456')
+                to='+123',
+                body=getattr(settings, 'TWILIO_MESSAGE').format(code),
+                from_='+456'
+            )
```
Comment (moggers87) [inline tests/test_gateways.py:69]: Would be nice to have a test with the default string too

File: two_factor/gateways/twilio/gateway.py [hunk 2]
```diff
@@ -52,7 +53,14 @@ def make_call(self, device, token):
...
 
     def send_sms(self, device, token):
-        body = gettext('Your authentication token is %s') % token
+        """
+        send sms using custom TWILIO_MESSAGE from settings.py
+        """
+        body = getattr(
+            settings,
+            'TWILIO_MESSAGE',
+            'Your authentication token is {}'
```
Comment (moggers87) [inline two_factor/gateways/twilio/gateway.py:62]: I think the default value should still be a translated string. For most people this has been fine and they might be a little annoyed when their users start getting English text messages

File: two_factor/gateways/twilio/gateway.py [hunk 3]
```diff
@@ -52,7 +53,14 @@ def make_call(self, device, token):
                                  url=uri, method='GET', timeout=15)
 
     def send_sms(self, device, token):
-        body = gettext('Your authentication token is %s') % token
+        """
+        send sms using custom TWILIO_MESSAGE from settings.py
+        """
+        body = getattr(
+            settings,
+            'TWILIO_MESSAGE',
```
Comment (moggers87) [inline two_factor/gateways/twilio/gateway.py:61]: I think a better name for this setting would be `TWO_FACTOR_SMS_MESSAGE`. It should also be documented so users know they can change it.

**Review by moggers87 — CHANGES_REQUESTED** [review]
Overall this looks good :+1:

There's just a few minor changes I'd like to see before merging.

**Comment by claudep** [comment]
An alternative would be to use a template, to not add yet another setting. Less straightforward to configure than a setting, but a lot more powerful as you can pass more context information.

**Comment by moggers87** [comment]
@claudep yet another template though :stuck_out_tongue: but I agree it's a better solution.

**Comment by aseem-hegshetye** [comment]
I have updated the PR with everything requested. Please let me know if more changes are required

File: two_factor/gateways/twilio/gateway.py [hunk 4]
```diff
@@ -51,8 +53,15 @@ def make_call(self, device, token):
                                  from_=getattr(settings, 'TWILIO_CALLER_ID'),
                                  url=uri, method='GET', timeout=15)
 
-    def send_sms(self, device, token):
-        body = gettext('Your authentication token is %s') % token
+    def send_sms(self, device, token, sms_message):
```
Comment (moggers87) [inline two_factor/gateways/twilio/gateway.py:56]: This method shouldn't require any extra parameters.

File: two_factor/templates/two_factor/twilio/sms_message.html [hunk 5]
```diff
@@ -0,0 +1,3 @@
+{% load i18n %}
+{% trans sms_message %}
```
Comment (moggers87) [inline two_factor/templates/two_factor/twilio/sms_message.html:2]: This doesn't need to be a variable, if users want to override the default message they will override the template itself.

File: two_factor/templates/two_factor/twilio/sms_message.html [hunk 6]
```diff
@@ -0,0 +1,3 @@
...
+{% trans 'Your token is '%}{{ token }}
```
Comment (moggers87) [inline two_factor/templates/two_factor/twilio/sms_message.html:3]: This should use `blocktrans` so translators have some context. Also the file should end with a newline.

**Review by moggers87 — CHANGES_REQUESTED** [review]
I'm not actually sure why tests are passing (unless I'm missing something, your changes will break sending SMS). That's not something that should be fixed in this PR, but it's something to be aware of :cat:

File: two_factor/templates/two_factor/twilio/sms_message.html (hunk 6 above)
Comment (aseem-hegshetye) [reply two_factor/templates/two_factor/twilio/sms_message.html]: done

File: two_factor/gateways/twilio/gateway.py (hunk 4 above)
Comment (aseem-hegshetye) [reply two_factor/gateways/twilio/gateway.py]: done

File: two_factor/templates/two_factor/twilio/sms_message.html (hunk 5 above)
Comment (aseem-hegshetye) [reply two_factor/templates/two_factor/twilio/sms_message.html]: done

**Comment by aseem-hegshetye** [comment]
Anything else I need to do for this PR?

**Comment by moggers87** [comment]
I need to re-review it and either merge or ask for more changes

File: tests/test_gateways.py [hunk 7]
```diff
@@ -98,6 +109,35 @@ def test_invalid_twilio_language(self, client):
             with translation.override('ar'):
                 twilio.make_call(device=Mock(number='+123'), token='654321')
 
+    @override_settings(
+        TWILIO_ACCOUNT_SID='SID',
+        TWILIO_AUTH_TOKEN='TOKEN',
+        TWILIO_CALLER_ID='+456',
+    )
+    @patch('two_factor.gateways.twilio.gateway.Client')
+    def test_twilio_blank_sms_message(self, client):
```
Comment (moggers87) [inline tests/test_gateways.py:118]: I think we can remove this test as we're now using a templates.

File: two_factor/templates/two_factor/twilio/sms_message.html [hunk 8]
```diff
@@ -0,0 +1,5 @@
+{% load i18n %}
+{% blocktrans %}
```
Comment (moggers87) [inline two_factor/templates/two_factor/twilio/sms_message.html:2]: `{% blocktrans trimmed %}` is better for translators (it removes unnecessary whitespace)

**Review by moggers87 — CHANGES_REQUESTED** [review]
Two more minor tweaks and then I think this is ready to merge :+1:

**Comment by aseem-hegshetye** [comment]
I have made all requested changes

**Review by moggers87 — APPROVED** [review]

**Comment by moggers87** [comment]
@aseem-hegshetye thanks for working on this :smiley_cat:
